torch/_inductor/subgraph_lowering.py

```diff
@@ -27,46 +41,89 @@ class PointwiseSubgraphLowering(torch.fx.Interpreter):
     """
 
     graph_outputs: Optional[List[ir.IRNode]]
+    root_graph: torch._inductor.graph.GraphLowering
+    # For backwards of buffer_grads with scatters we allow mutations
+    allowed_mutations: Optional[Set[OpOverload]] = None
+    additional_lowerings: Optional[LoweringDict] = None
+    _current_op: Optional[TargetType] = None
 
     def __init__(
         self,
         gm: torch.fx.GraphModule,
-        root_graph_lowering: "torch._inductor.graph.GraphLowering",
+        root_graph_lowering: torch._inductor.graph.GraphLowering,
+        allowed_mutations: Optional[Set[OpOverload]] = None,
+        additional_lowerings: Optional[LoweringDict] = None,
     ) -> None:
         super().__init__(gm)
         self.graph_outputs = None
         self.root_graph = root_graph_lowering
+        self.allowed_mutations = allowed_mutations
+        self.additional_lowerings = additional_lowerings
+        self._current_op = None
+
+    @contextmanager
+    def _op_context(self, op: TargetType) -> Generator[None, None, None]:
+        """Set which op is being processed in call function to know if we can mutate buffers"""
+        previous = self._current_op
+        self._current_op = op
+        try:
+            yield
+        finally:
+            self._current_op = previous
+
+    def _approved_mutator(self) -> bool:
+        return (
+            self.allowed_mutations is not None
+            and self._current_op in self.allowed_mutations
+        )
 
     def mark_buffer_mutated(self, name: str) -> None:
-        raise SubgraphLoweringException("Mutations are not supported in this context")
+        if self._approved_mutator():
+            return self.root_graph.mark_buffer_mutated(name)
+        else:
+            raise SubgraphLoweringException(
+                f"Buffer mutation of '{name}' detected during lowering of {self._current_op}. "
+                "Buffer mutations are only allowed in approved mutation ops. "
+                "This is an error in the lowering of the subgraph, please file a bug report."
+            )
 
-    def register_buffer(self, buffer: ir.Buffer) -> str:
-        raise SubgraphLoweringException(
-            "Buffers cannot be created while lowering a pointwise subgraph. "
-            "This could be for a good reason (e.g. you're calling an op we can't codegen as a pointwise op), "
-            "but it could also be a bug. Please file a bug report if you think this should be supportable."
-        )
+    def register_buffer(self, buffer: ir.Buffer, *, set_name: bool = False) -> str:
+        if self._approved_mutator():
+            return self.root_graph.register_buffer(buffer, set_name=set_name)
+        else:
+            raise SubgraphLoweringException(
+                f"Buffer '{buffer.name}' cannot be registered during the lowering of {self._current_op}. "
```

Comment:
This error message would be more helpful, but because Reduction operations are not necessarily the target that registers the buffer. The error message can be kind of confusing. For instance, for the existing captured buffer test, this is what we see.

targets that get looped over:
```
aten.sum.dim_IntList
aten.sqrt.default
aten.index.Tensor
```

`Buffers cannot be registered during the lowering of aten.index.Tensor. `

So although there is a reduction which breaks the pointwise lowering i.e the sum.dim list it's not until we actually access the elements that we have to register the new buffer. So maybe we just remove the current op from this error string since it can be kind of confusing.